docs/EVALUATION.md

```diff
@@ -0,0 +1,249 @@
+
+## Setting
+
+We are looking at a completed pipeline run that produced multiple triggers and therefore multiple models. We want to evaluate the performance of each model.
+
+```mermaid
+gantt
+    title Pipeline run
+    dateFormat YYYY-MM-DD
+    axisFormat %Y
+    section models
+        Trigger1 :milestone, 1930, 0d
+        model 1          :a1, 1930, 10y
+        Trigger2 :milestone, :after a1
+        model 2          :a2, after a1, 5y
+        Trigger3 :milestone, :after a2
+        model 3          :a3, after a2, 5y
+        Trigger4 :milestone, :after a3
+        model 4          :a4, after a3, 25y
+        Trigger5 :milestone, :after a4
+        model 5          :a5, after a4, 25y
+        Trigger6 :milestone, :after a5
+        model 6          :a6, after a5, 5y
+        Trigger7 :milestone, :after a6
+        model 7          :a7, after a6, 5y
+```
+
+## Evaluation methodology
+
+An evaluation methodology consists of three main components:
+
+1) The evaluation metric (e.g. F1, accuracy, precision, recall, etc.)
+2) The evaluation data (set of samples out of the evaluation split used for single model evaluation)
+3) The points in time when the evaluations are run
+
+### Use case 1: online evaluation
+
+In this case we have to be extremely cautious to keep evaluation costs low as the are part of the trigger policy evaluation costs (or at least part of the core pipeline that cannot be disregarded like the experiment model evaluations).
+
+During the core pipeline execution (not including experiment mode pipeline steps), we cannot access future data for evaluation. Also we don't have an evaluation split available. Therefore, we need to use the training data up to the current point in time for evaluation. This is not ideal but it is the only way to evaluate the model in an online fashion.
+
+> The question of how far into the past we want to includes samples remains. More thought is needed before continuing on this.
+
+### Use case 2: offline evaluation
+
+In experiment mode we can potentially utilize the full evaluation split over the whole dataset time range. Being able to "look into the future" is important to determine the performance of a model.
+
+#### [2] Total pipeline performance
+
+We argue that the performance of pipeline w.r.t. a certain evaluation metric is the weighted average of the performance of the most up to date model at each point in time over the whole time range of the pipeline run.
+
+Mind, that we need to weight every model performance at a point in time by the number of samples within that time range.
+
+> E.g. in `model 3` from the Gantt diagram above: assuming our dataset sample distribution is so uneven that 90% of the samples are in the time range between `Trigger3` and `Trigger4`, then the performance of `model 3` should have a higher weight in the pipeline performance than the performance of `model 2`.
```

Comment:
I am not 100% sure yet we need to weight by number of samples used for training. Just because we have more samples to train a model does not necessarily imply for me its performance is more important. I'd argue I am always equally interested that at whatever timestep I use the current model, it performs well.

Reply:
Okay, what I intended to do there is to not assign the same weights to a model that was used for 1 day to one that was used for 1 year (without retrainings). Using weighting by the time a model was actively used, captures this idea better!